runtime/src/accounts_index.rs

```diff
@@ -181,9 +208,198 @@ impl<'a, T: 'static + Clone> Iterator for AccountsIndexIterator<'a, T> {
+
+                    // Write lock on `inner_key_entry` above through the `entry`
+                    // means nobody else has access to this lock at this time,
+                    // so this check for empty -> remove() is atomic
+                    if slot_set.read().unwrap().is_empty() {
+                        inner_key_entry.remove();
+                    }
+                }
+                inner_key_map.is_empty()
+            } else {
+                false
+            };
+            is_key_empty
+        } else {
+            false
+        };
+
+        // Delete the `key` if the set of inner keys is empty
+        if is_key_empty {
+            if let dashmap::mapref::entry::Entry::Occupied(key_entry) = self.index.entry(*key) {
+                if key_entry.get().is_empty() {
+                    key_entry.remove();
+                }
+            }
+        }
+    }
+
+    // Specifying `slots_to_remove` == None will only remove keys for those specific slots
+    // found for the `inner_key` in the reverse index. Otherwise, passing `None`
+    // will  remove all keys that are found for the `inner_key` in the reverse index.
+
+    // Note passing `None` is dangerous unless you're sure there's no other competing threads
+    // writing updates to the index for this Pubkey at the same time!
+    fn remove_by_inner_key(&self, inner_key: &Pubkey, slots_to_remove: Option<&HashSet<Slot>>) {
+        // Save off which keys in `self.index` had slots removed so we can remove them
+        // after we purge the reverse index
+        let mut key_to_removed_slots: HashMap<Pubkey, Vec<Slot>> = HashMap::new();
+
+        // Check if the entry for `inner_key` in the reverse index is empty
+        // and can be removed
+        let needs_remove = {
+            if let Some(slots_to_remove) = slots_to_remove {
+                self.reverse_index
+                    .get(inner_key)
+                    .map(|slots_map| {
+                        // Ideally we use a concurrent map here as well to prevent clean
+                        // from blocking writes, but memory usage of DashMap is high
+                        let mut w_slots_map = slots_map.value().write().unwrap();
+                        for slot in slots_to_remove.iter() {
+                            if let Some(removed_key) = w_slots_map.remove(slot) {
+                                key_to_removed_slots
+                                    .entry(removed_key)
+                                    .or_default()
+                                    .push(*slot);
+                            }
+                        }
+                        w_slots_map.is_empty()
+                    })
+                    .unwrap_or(false)
+            } else {
+                if let Some((_, removed_slot_map)) = self.reverse_index.remove(inner_key) {
+                    for (slot, key) in removed_slot_map.into_inner().unwrap().into_iter() {
```

Comment:
nits: rename to `removed_key` to align with the code in the above `if`-then clause.